File: 0575. Distribute Candies/0575.java
// time complexity: O(nlogn)
// space complexity: O(1)
class Solution {
    public int distributeCandies(int[] candyType) {
        // n is even number
        int n = candyType.length;

        // if there are as many different numbers as n/2, return n/2.
        // if there are x different numbers where is x < n/2, return x.
        // if x > n/2, return n/2.

        Arrays.sort(candyType);
        int different=1;
        for(int i=1; i<n; i++){
            if(candyType[i] != candyType[i-1]) different++;
        }

        // return different < n/2 ? different : n/2;
        return Math.min(n/2, different);
    }
}

File: 0575. Distribute Candies/0575.py
# time complexity: O(n)
# space complexity: O(n) - (worst case)
class Solution(object):
    def distributeCandies(self, candyType):
        """
        :type candyType: List[int]
        :rtype: int
        """
        # unique number of candies
        unique_number = len(set(candyType))
        # max number of candies that can be eaten
        n = len(candyType) // 2

        return min(n, unique_number)
